[ICE]:  error performing operation: query type op

<!--
[31mICE[0m: Rustc ./snippet_0.rs '-Znext-solver=globally --crate-type=lib -ooutputfile -Zdump-mir-dir=dir' 'error: internal compiler error: error performing operation: query type op', 'error: internal compiler error: error performing operation: query type op'
File: /tmp/im/snippet_0.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -Znext-solver=globally --crate-type=lib
pub trait Trait<'a> {
    type Type;
}

pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}

````

original:
````rust
pub trait Trait<'a> {
    type Type;
}

pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {
    
}
````

Version information
````
rustc 1.95.0-nightly (621d76794 2026-01-18)
binary: rustc
commit-hash: 621d76794c76fc21c0a6151fbc090120e0230a91
commit-date: 2026-01-18
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

````



Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Znext-solver=globally --crate-type=lib`

<details><summary><strong>Program output</strong></summary>
<p>

```
warning: unused variable: `v`
 --> /tmp/icemaker_global_tempdir.VllVRX9PaHyX/rustc_testrunner_tmpdir_reporting.n5W4SveOCKsj/mvce.rs:5:44
  |
5 | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
  |                                            ^ help: if this is intentional, prefix it with an underscore: `_v`
  |
  = note: `#[warn(unused_variables)]` (part of `#[warn(unused)]`) on by default

warning: 1 warning emitted

note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error: error performing operation: query type op
 --> /tmp/icemaker_global_tempdir.VllVRX9PaHyX/rustc_testrunner_tmpdir_reporting.n5W4SveOCKsj/mvce.rs:5:1
  |
5 | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: delayed at /rustc-dev/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_trait_selection/src/traits/query/type_op/custom.rs:95:25
         0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
         1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
         2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
         3: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, alloc::string::String>
         4: <rustc_borrowck::type_check::free_region_relations::UniversalRegionRelationsBuilder>::add_implied_bounds
         5: rustc_borrowck::type_check::type_check
         6: <rustc_borrowck::root_cx::BorrowCheckRootCtxt>::do_mir_borrowck
         7: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 8]>>
         8: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>, rustc_query_system::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
         9: rustc_query_impl::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
        10: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
        11: rustc_interface::passes::analysis
        12: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
        13: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
        14: rustc_query_impl::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
        15: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
        16: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
        17: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
        18: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        19: <std::sys::thread::unix::Thread>::new::thread_start
        20: <unknown>
        21: <unknown>
      
 --> /tmp/icemaker_global_tempdir.VllVRX9PaHyX/rustc_testrunner_tmpdir_reporting.n5W4SveOCKsj/mvce.rs:5:1
  |
5 | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error: error performing operation: query type op
 --> /tmp/icemaker_global_tempdir.VllVRX9PaHyX/rustc_testrunner_tmpdir_reporting.n5W4SveOCKsj/mvce.rs:5:1
  |
5 | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: delayed at /rustc-dev/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_trait_selection/src/traits/query/type_op/custom.rs:95:25
         0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
         1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
         2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
         3: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, alloc::string::String>
         4: rustc_borrowck::type_check::type_check
         5: <rustc_borrowck::root_cx::BorrowCheckRootCtxt>::do_mir_borrowck
         6: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 8]>>
         7: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>, rustc_query_system::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
         8: rustc_query_impl::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
         9: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
        10: rustc_interface::passes::analysis
        11: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
        12: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
        13: rustc_query_impl::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
        14: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
        15: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
        16: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
        17: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
        18: <std::sys::thread::unix::Thread>::new::thread_start
        19: <unknown>
        20: <unknown>
      
 --> /tmp/icemaker_global_tempdir.VllVRX9PaHyX/rustc_testrunner_tmpdir_reporting.n5W4SveOCKsj/mvce.rs:5:1
  |
5 | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error: error performing operation: query type op
  |
  = note: delayed at /rustc-dev/621d76794c76fc21c0a6151fbc090120e0230a91/compiler/rustc_trait_selection/src/traits/query/type_op/custom.rs:95:25
             0: <rustc_errors::DiagCtxtInner>::emit_diagnostic
             1: <rustc_errors::DiagCtxtHandle>::emit_diagnostic
             2: <rustc_span::ErrorGuaranteed as rustc_errors::diagnostic::EmissionGuarantee>::emit_producing_guarantee
             3: <rustc_errors::DiagCtxtHandle>::span_delayed_bug::<rustc_span::span_encoding::Span, alloc::string::String>
             4: <rustc_borrowck::type_check::liveness::trace::LivenessContext>::compute_drop_data
             5: <rustc_borrowck::type_check::liveness::trace::LivenessContext>::add_drop_live_facts_for
             6: rustc_borrowck::type_check::liveness::generate
             7: rustc_borrowck::type_check::type_check
             8: <rustc_borrowck::root_cx::BorrowCheckRootCtxt>::do_mir_borrowck
             9: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 8]>>
            10: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_data_structures::vec_cache::VecCache<rustc_span::def_id::LocalDefId, rustc_middle::query::erase::Erased<[u8; 8]>, rustc_query_system::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
            11: rustc_query_impl::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
            12: <rustc_middle::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
            13: rustc_interface::passes::analysis
            14: rustc_query_impl::plumbing::__rust_begin_short_backtrace::<rustc_query_impl::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle::query::erase::Erased<[u8; 0]>>
            15: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::DynamicConfig<rustc_query_system::query::caches::SingleCache<rustc_middle::query::erase::Erased<[u8; 0]>>, false, false, false>, rustc_query_impl::plumbing::QueryCtxt, false>
            16: rustc_query_impl::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
            17: <rustc_interface::passes::create_and_enter_global_ctxt<core::option::Option<rustc_interface::queries::Linker>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core::ops::function::FnOnce<(&rustc_session::session::Session, rustc_middle::ty::context::CurrentGcx, alloc::sync::Arc<rustc_data_structures::jobserver::Proxy>, &std::sync::once_lock::OnceLock<rustc_middle::ty::context::GlobalCtxt>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_middle::arena::Arena>, &rustc_data_structures::sync::worker_local::WorkerLocal<rustc_hir::Arena>, rustc_driver_impl::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
            18: rustc_interface::interface::run_compiler::<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}
            19: std::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
            20: <std::thread::lifecycle::spawn_unchecked<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<(), rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
            21: <std::sys::thread::unix::Thread>::new::thread_start
            22: <unknown>
            23: <unknown>
          

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.95.0-nightly (621d76794 2026-01-18) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally --crate-type lib -Z dump-mir-dir=dir

query stack during panic:
end of query stack

```

</p>
</details>

<!--
query stack:

-->
@rustbot label +WG-trait-system-refactor

old solver:
```
error[E0284]: type annotations needed
 --> /tmp/crash.rs:5:47
  |
5 | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
  |                                               ^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
  |
  = note: cannot satisfy `<T as Trait<'a>>::Type == _`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0284`.
```

--- a/compiler/rustc_trait_selection/src/traits/query/type_op/custom.rs
+++ b/compiler/rustc_trait_selection/src/traits/query/type_op/custom.rs
@@ -92,7 +92,40 @@ where
     let value = infcx.commit_if_ok(|_| {
         let ocx = ObligationCtxt::new(infcx);
         let value = op(&ocx).map_err(|_| {
-            infcx.dcx().span_delayed_bug(span, format!("error performing operation: {name}"))
+            infcx.tcx.check_potentially_region_dependent_goals(root_def_id).err().unwrap_or_else(
+                // FIXME: In this region-dependent context, `type_op` should only fail due to
+                // region-dependent goals. Any other kind of failure indicates a bug and we
+                // should ICE.
+                //
+                // In principle, all non-region errors are expected to be emitted before
+                // borrowck. Such errors should taint the body and prevent borrowck from
+                // running at all. However, this invariant does not currently hold.
+                //
+                // Consider:
+                //
+                // ```ignore (illustrative)
+                // struct Foo<T>(T)
+                // where
+                //     T: Iterator,
+                //     <T as Iterator>::Item: Default;
+                //
+                // fn foo<T>() -> Foo<T> {
+                //     loop {}
+                // }
+                // ```
+                //
+                // Here, `fn foo` ought to error and taint the body before MIR build, since
+                // `Foo<T>` is not well-formed. However, we currently avoid checking this
+                // during HIR typeck to prevent duplicate diagnostics.
+                //
+                // If we ICE here on any non-region-dependent failure, we would trigger ICEs
+                // too often for such cases. For now, we emit a delayed bug instead.
+                || {
+                    infcx
+                        .dcx()
+                        .span_delayed_bug(span, format!("error performing operation: {name}"))
+                },
+            )
         })?;
         let errors = ocx.evaluate_obligations_error_on_ambiguity();
         if errors.is_empty() {

--- a/tests/ui/traits/next-solver/assembly/ambiguity-due-to-uniquification-4.current.stderr
+++ b/tests/ui/traits/next-solver/assembly/ambiguity-due-to-uniquification-4.current.stderr
@@ -1,0 +1,20 @@
+error[E0284]: type annotations needed
+  --> $DIR/ambiguity-due-to-uniquification-4.rs:17:47
+   |
+LL | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
+   |                                               ^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
+   |
+   = note: cannot satisfy `<T as Trait<'a>>::Type == _`
+
+error[E0284]: type annotations needed
+  --> $DIR/ambiguity-due-to-uniquification-4.rs:17:47
+   |
+LL | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
+   |                                               ^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
+   |
+   = note: cannot satisfy `<T as Trait<'a>>::Type == _`
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0284`.

--- a/tests/ui/traits/next-solver/assembly/ambiguity-due-to-uniquification-4.next.stderr
+++ b/tests/ui/traits/next-solver/assembly/ambiguity-due-to-uniquification-4.next.stderr
@@ -1,0 +1,9 @@
+error[E0284]: type annotations needed: cannot normalize `<T as Trait<'_>>::Type`
+  --> $DIR/ambiguity-due-to-uniquification-4.rs:17:44
+   |
+LL | pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
+   |                                            ^ cannot normalize `<T as Trait<'_>>::Type`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0284`.

--- a/tests/ui/traits/next-solver/assembly/ambiguity-due-to-uniquification-4.rs
+++ b/tests/ui/traits/next-solver/assembly/ambiguity-due-to-uniquification-4.rs
@@ -1,0 +1,21 @@
+//@ revisions: current next
+//@[next] compile-flags: -Znext-solver
+//@ ignore-compare-mode-next-solver (explicit revisions)
+
+// A regression test for https://github.com/rust-lang/rust/issues/151318.
+//
+// Unlike in the previous other tests, this fails to compile with the old solver as well.
+// Although we were already stashing goals which depend on inference variables and then
+// reproving them at the end of HIR typeck to avoid causing an ICE during MIR borrowck,
+// it wasn't enough because the type op itself can result in an error due to uniquification,
+// e.g. while normalizing a projection type.
+
+pub trait Trait<'a> {
+    type Type;
+}
+
+pub fn f<'a, 'b, T: Trait<'a> + Trait<'b>>(v: <T as Trait<'a>>::Type) {}
+//~^ ERROR type annotations needed
+//[current]~| ERROR type annotations needed
+
+fn main() {}

--- a/tests/ui/wf/return-type-non-wf-no-ice.current.stderr
+++ b/tests/ui/wf/return-type-non-wf-no-ice.current.stderr
@@ -1,0 +1,22 @@
+error[E0277]: `T` is not an iterator
+  --> $DIR/return-type-non-wf-no-ice.rs:13:16
+   |
+LL | fn foo<T>() -> Foo<T> {
+   |                ^^^^^^ `T` is not an iterator
+   |
+note: required by a bound in `Foo`
+  --> $DIR/return-type-non-wf-no-ice.rs:10:8
+   |
+LL | pub struct Foo<T>(T)
+   |            --- required by a bound in this struct
+LL | where
+LL |     T: Iterator,
+   |        ^^^^^^^^ required by this bound in `Foo`
+help: consider restricting type parameter `T` with trait `Iterator`
+   |
+LL | fn foo<T: std::iter::Iterator>() -> Foo<T> {
+   |         +++++++++++++++++++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/wf/return-type-non-wf-no-ice.next.stderr
+++ b/tests/ui/wf/return-type-non-wf-no-ice.next.stderr
@@ -1,0 +1,22 @@
+error[E0277]: `T` is not an iterator
+  --> $DIR/return-type-non-wf-no-ice.rs:13:16
+   |
+LL | fn foo<T>() -> Foo<T> {
+   |                ^^^^^^ `T` is not an iterator
+   |
+note: required by a bound in `Foo`
+  --> $DIR/return-type-non-wf-no-ice.rs:10:8
+   |
+LL | pub struct Foo<T>(T)
+   |            --- required by a bound in this struct
+LL | where
+LL |     T: Iterator,
+   |        ^^^^^^^^ required by this bound in `Foo`
+help: consider restricting type parameter `T` with trait `Iterator`
+   |
+LL | fn foo<T: std::iter::Iterator>() -> Foo<T> {
+   |         +++++++++++++++++++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/tests/ui/wf/return-type-non-wf-no-ice.rs
+++ b/tests/ui/wf/return-type-non-wf-no-ice.rs
@@ -1,0 +1,18 @@
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@ [next] compile-flags: -Znext-solver
+
+// Ensure we do not ICE if a non-well-formed return type manages to slip past HIR typeck.
+// See <https://github.com/rust-lang/rust/pull/152816> for details.
+
+pub struct Foo<T>(T)
+where
+    T: Iterator,
+    <T as Iterator>::Item: Default;
+
+fn foo<T>() -> Foo<T> {
+    //~^ ERROR: `T` is not an iterator
+    loop {}
+}
+
+fn main() {}

